### PR TITLE
Replace xpath

### PR DESCRIPTION
**Pull request by kwkwan**

### CODE REVIEW

**Comment by kwkwan** [comment]
As parsing xmi file by `lutaml/xmi` is still NOT fully work at this moment, the commit works are not tested.

Some of the `xpath` are not converted.  Unlike navigating elements by ruby object, `xpath` can be powerful to go to deep inside in different branches in a tree.  The proper way to handle these situation are still investigating.

File: lib/lutaml/xmi/parsers/xml.rb [hunk 1]
```diff
@@ -12,39 +13,44 @@ class XML
...
-        # @return [Lutaml::XMI::Model::Document]
-        def self.parse(io, _options = {})
-          new.parse(Nokogiri::XML(io.read))
+        attr_reader :main_model, :xmi_cache, :xmi_root_model
+
+        # @param xml [String] path to xml
+        # @param options [Hash] options for parsing
+        # @return [Lutaml::Uml::Document]
+        def self.parse(xml, _options = {})
+          xml_content = File.read(xml).gsub("\t", "  ")
```
Comment (ronaldtse) [inline lib/lutaml/xmi/parsers/xml.rb:22]: We don’t need to remove Tabs because we are not doing a compare.

File: lib/lutaml/xmi/parsers/xml.rb [hunk 2]
```diff
@@ -55,53 +61,52 @@ def serialize_model_packages(model)
...
-              constraints: serialize_class_constraints(klass),
-              is_abstract: doc_node_attribute_value(klass, "isAbstract"),
-              definition: doc_node_attribute_value(klass, "documentation"),
-              stereotype: doc_node_attribute_value(klass, "stereotype")
-            }
+        # @note xpath ./packagedElement[@xmi:type="uml:Class" or
+        #                               @xmi:type="uml:AssociationClass"]
+        def serialize_model_classes(package, model)
+          package.packaged_element.select { |e|
+            e.type == "uml:Class" || e.type == "uml:AssociationClass"
```
Comment (ronaldtse) [inline lib/lutaml/xmi/parsers/xml.rb:68]: Can we do “is_a?(Uml::Class)” something like that?

File: lib/lutaml/xmi/parsers/xml.rb [hunk 3]
```diff
@@ -55,53 +61,52 @@ def serialize_model_packages(model)
...
-                type = value.xpath(".//type").first || {}
-                {
-                  name: value["name"],
-                  type: lookup_entity_name(type["xmi:idref"]) || type["xmi:idref"],
-                  definition: lookup_attribute_definition(value),
-                }
+        # @note xpath ./packagedElement[@xmi:type="uml:Enumeration"]
+        def serialize_model_enums(package)
+          package.packaged_element
+            .select { |e| e.type == "uml:Enumeration" }.map do |enum|
```
Comment (ronaldtse) [inline lib/lutaml/xmi/parsers/xml.rb:89]: Same here, can we use “is_a?”?

File: lib/lutaml/xmi/parsers/xml.rb [hunk 4]
```diff
@@ -213,110 +220,145 @@ def serialize_member_type(owner_xmi_id, link, link_member_name)
...
+          member_end_cardinality, member_end_attribute_name =
+            fetch_owned_attribute_node(xmi_id)
 
           [member_end, member_end_type, member_end_cardinality, nil, xmi_id]
         end
 
+        # @note xpath
+        #   %(//ownedAttribute[@association]/type[@xmi:idref="#{xmi_id}"])
+        def fetch_owned_attribute_node(xmi_id)
+          node = main_model.xpath(%(//ownedAttribute[@association]/type[@xmi:idref="#{xmi_id}"])).first
```
Comment (ronaldtse) [inline lib/lutaml/xmi/parsers/xml.rb:293]: Will we replace this later?

File: lib/lutaml/xmi/parsers/xml.rb [hunk 5]
```diff
@@ -213,110 +220,145 @@ def serialize_member_type(owner_xmi_id, link, link_member_name)
...
-                # xmi_id: klass['xmi:id'],
-                name: attribute["name"],
-                type: lookup_entity_name(type["xmi:idref"]) || type["xmi:idref"],
-                xmi_id: type["xmi:idref"],
-                is_derived: attribute["isDerived"],
-                cardinality: { "min" => cardinality_min_value(attribute), "max" => cardinality_max_value(attribute) },
-                definition: lookup_attribute_definition(attribute),
-              }
-            end
+          klass.owned_attribute.select { |attr| attr.type == "uml:Property" }
```
Comment (ronaldtse) [inline lib/lutaml/xmi/parsers/xml.rb:320]: Can we use “is_a?”?

File: lib/lutaml/xmi/parsers/xml.rb [hunk 6]
```diff
@@ -213,110 +220,145 @@ def serialize_member_type(owner_xmi_id, link, link_member_name)
...
           doc_node = main_model.xpath(%(//element[@xmi:idref="#{xmi_id}"]/properties)).first
           return unless doc_node
 
           doc_node.attributes[attr_name]&.value
         end
 
-        def lookup_attribute_definition(node)
-          xmi_id = node["xmi:id"]
+        def lookup_attribute_documentation(xmi_id)
           doc_node = main_model.xpath(%(//attribute[@xmi:idref="#{xmi_id}"]/documentation)).first
```
Comment (ronaldtse) [inline lib/lutaml/xmi/parsers/xml.rb:362]: We’ll replace this later?

**Review by ronaldtse — CHANGES_REQUESTED** [review]
Very happy with the changes so far!

File: lib/lutaml/xmi/parsers/xml.rb (hunk 1 above)
Comment (kwkwan) [reply lib/lutaml/xmi/parsers/xml.rb]: Fixed in 3fc1ecea4aa1ae2d1bf4752baca8cabe0f73d71b.

File: lib/lutaml/xmi/parsers/xml.rb (hunk 4 above)
Comment (kwkwan) [reply lib/lutaml/xmi/parsers/xml.rb:293]: Yes, it should be replaced later.

File: lib/lutaml/xmi/parsers/xml.rb (hunk 6 above)
Comment (kwkwan) [reply lib/lutaml/xmi/parsers/xml.rb:362]: Yes, it should be replaced later.

File: lib/lutaml/xmi/parsers/xml.rb (hunk 2 above)
Comment (kwkwan) [reply lib/lutaml/xmi/parsers/xml.rb:68]: Will implement `is_type?` in `Shale::Mapper` class to handle these issues.

Comment (ronaldtse) [reply lib/lutaml/xmi/parsers/xml.rb]: We don’t want new methods? Aren’t AssociatedClass etc separate classes?

**Pull request merged**